apps/indexers/nft-marketplace/src/handlers/marketplace.ts

```diff
@@ -0,0 +1,168 @@
+      ? await store.findOneBy(NftContract, { id: innerPayload.nftContractId })
+      : undefined;
+
+  const token =
+    typeof innerPayload === 'object' && 'nftContractId' in innerPayload && 'tokenId' in innerPayload
+      ? await store.findOne(Token, {
+          where: { id: `${innerPayload.nftContractId}-${innerPayload.tokenId}` },
+          relations: { owner: true },
+        })
+      : undefined;
+
+  const variants: Record<string, Function> = {
+    nftContractAdded: async (id: NftId) => {
+      await nftState.add(id, store);
+      actionLog(`NEW NFT_CONTRACT`, id);
+      await indexNft(id, config.squid.startBlock, height - 1, store);
+    },
+    marketDataAdded: async ({ price }: IMarketPlaceDataAdded) => {
+      token.isListed = true;
+      if (price != null) {
+        token.price = BigInt(price);
+      }
+      await store.save(token);
+      actionLog(`MARKET DATA ADDED`, token.id);
+    },
+    itemSold: async ({ price }: IItemSold) => {
+      const transfer = new Transfer({
+        id,
+        token,
+        from: token.owner,
+        to: await getAcc(message.destination, store),
+        timestamp: BigInt(timestamp),
+        price: BigInt(price),
+      });
+
+      // token.transfers.push(transfer);
+      token.price = null;
+      await store.save(token);
+      await store.save(transfer);
+      actionLog(`ITEM SOLD`, `${token.id}, price: ${price}`);
+    },
+    auctionCreated: async ({ ftContractId, price }: IAuctionCreated) => {
+      const auction = new Auction({
+        id,
+        nftContract,
+        token,
+        ftContract: ftContractId,
+        finishAt: BigInt(timestamp) + BigInt(0),
+        price: BigInt(price),
+        bids: [],
+        isOpened: true,
+      });
+      await store.save(auction);
+      actionLog(`AUCTION CREATED`, token.id);
+    },
+    bidAdded: async ({ price }: IBidAdded) => {
+      const auction = await store.findOneBy(Auction, { nftContract });
+      const bid = new Bid({ id, auction, price: BigInt(price) });
+      auction.price = BigInt(price);
+      // auction.bids.push(bid);
+      await store.save(auction);
+      await store.save(bid);
+      actionLog(`BID ADDED`, `${token.id}, price: ${price}`);
+    },
+    auctionSettled: async ({ price, newOwner }: IAuctionSettled) => {
+      const auction = await store.findOneBy(Auction, { nftContract });
+      auction.price = BigInt(price);
+      auction.isOpened = false;
+      const owner = await store.findOneBy(Account, { id: newOwner });
+      const transfer = new Transfer({
+        id,
+        token,
+        from: token.owner,
+        to: owner,
+        timestamp: BigInt(timestamp),
+        price: BigInt(price),
+      });
+      token.owner = await store.findOneBy(Account, { id: newOwner });
+      // token.transfers.push(transfer);
+      token.price = null;
+      await store.save(token);
+      await store.save(auction);
+      await store.save(transfer);
```

Comment:
it is possible to resolve asynchronous processes with `Promise.all` here

And some others below?

Reply:
Database queries are executed one-by-one in a single transaction. So, `Promise.all` would just create additional promises slightly degrading performance.